File: problems/problemset/mini-redis/Python/solution/mini_redis.py
import socket
from collections import deque

WRONGTYPE = "WRONGTYPE Operation against a key holding the wrong kind of value"

class _Entry:
    __slots__ = ('type', 'str_val', 'list_val', 'set_val', 'zset_val', 'expire_at')
    def __init__(self, type_=None):
        self.type      = type_
        self.str_val   = ''
        self.list_val  = deque()
        self.set_val   = set()
        self.zset_val  = {}   # member -> score
        self.expire_at = -1

class _Store:
    def __init__(self):
        self._d   = {}
        self._now = 0

    def _dead(self, k):
        e = self._d.get(k)
        if e is None: return True
        if e.expire_at >= 0 and e.expire_at <= self._now:
            del self._d[k]; return True
        return False

    def alive(self, k):
        return not self._dead(k)

    def get(self, k):
        return None if self._dead(k) else self._d[k]

    def make(self, k, t):
        if self._dead(k):
            e = _Entry(t); self._d[k] = e
        else:
            e = self._d[k]
            if e.type is None: e.type = t
        return e

    def reset(self, k):
        e = _Entry(); self._d[k] = e; return e

    def erase(self, k):  self._d.pop(k, None)
    def flush(self):     self._d.clear()
    def tick(self, n):   self._now += n
    def time(self):      return self._now

    def dbsize(self):
        ks = list(self._d.keys())
        return sum(1 for k in ks if self.alive(k))


class MiniRedis:
    def __init__(self):
        self._s = _Store()

    # ── RESP reader ───────────────────────────────────────────────────────────

    @staticmethod
    def _read_cmd(buf, conn):
        # buf is a bytearray used as a read buffer
        def recv_into():
            chunk = conn.recv(4096)
            if chunk: buf.extend(chunk)
            return bool(chunk)

        def read_line():
            while True:
                nl = buf.find(b'\n')
                if nl >= 0:
                    line = bytes(buf[:nl+1])
                    del buf[:nl+1]
                    return line.rstrip(b'\r\n')
                if not recv_into(): return None

        def read_exact(n):
            while len(buf) < n + 2:
                if not recv_into(): return None
            data = bytes(buf[:n])
            del buf[:n+2]
            return data

        line = read_line()
        if not line or not line.startswith(b'*'): return None
        try: argc = int(line[1:])
        except ValueError: return None
        args = []
        for _ in range(argc):
            blen = read_line()
            if not blen or not blen.startswith(b'$'): return None
            try: n = int(blen[1:])
            except ValueError: return None
            arg = read_exact(n)
            if arg is None: return None
            args.append(arg.decode())
        return args

    # ── RESP writer ───────────────────────────────────────────────────────────

    @staticmethod
    def _w(conn, s): conn.sendall(s if isinstance(s, bytes) else s.encode())

    @classmethod
    def _simple(cls, c, s):   cls._w(c, f'+{s}\r\n')
    @classmethod
    def _err(cls, c, s):      cls._w(c, f'-{s}\r\n')
    @classmethod
    def _int(cls, c, n):      cls._w(c, f':{n}\r\n')
    @classmethod
    def _nil(cls, c):         cls._w(c, '$-1\r\n')
    @classmethod
    def _earr(cls, c):        cls._w(c, '*0\r\n')

    @classmethod
    def _bulk(cls, c, s):
        b = s.encode(); cls._w(c, f'${len(b)}\r\n'.encode() + b + b'\r\n')

    @classmethod
    def _arr(cls, c, items):
        cls._w(c, f'*{len(items)}\r\n')
        for item in items: cls._bulk(c, item)

    # ── Command dispatch ──────────────────────────────────────────────────────

    def _dispatch(self, conn, args):
        if not args: return
        cmd = args[0].upper()
        s = self._s

        if cmd == 'PING':
            if len(args) > 1: self._bulk(conn, args[1])
            else: self._simple(conn, 'PONG')

        elif cmd == 'FLUSHALL':
            s.flush(); self._simple(conn, 'OK')

        elif cmd == 'DBSIZE':
            self._int(conn, s.dbsize())

        elif cmd == 'DEL':
            n = 0
            for k in args[1:]:
                if s.alive(k): s.erase(k); n += 1
            self._int(conn, n)

        elif cmd == 'EXISTS':
            self._int(conn, sum(1 for k in args[1:] if s.alive(k)))

        elif cmd == 'TYPE':
            e = s.get(args[1])
            self._simple(conn, e.type if e else 'none')

        elif cmd == 'EXPIRE':
            e = s.get(args[1])
            if not e: self._int(conn, 0); return
            e.expire_at = s.time() + int(args[2])
            self._int(conn, 1)

        elif cmd == 'TTL':
            e = s.get(args[1])
            if not e: self._int(conn, -2); return
            self._int(conn, -1 if e.expire_at < 0 else e.expire_at - s.time())

        elif cmd == 'DEBUG':
            sub = args[1].upper() if len(args) > 1 else ''
            if sub == 'TICK' and len(args) >= 3:
                s.tick(int(args[2])); self._simple(conn, 'OK')
            else:
                self._err(conn, 'ERR unknown DEBUG subcommand')

        elif cmd == 'SET':
            e = s.reset(args[1]); e.type = 'string'; e.str_val = args[2]
            i = 3
            while i + 1 < len(args):
                if args[i].upper() == 'EX':
                    e.expire_at = s.time() + int(args[i+1]); i += 2
                else: i += 1
            self._simple(conn, 'OK')

        elif cmd == 'GET':
            e = s.get(args[1])
            if not e: self._nil(conn); return
            if e.type != 'string': self._err(conn, WRONGTYPE); return
            self._bulk(conn, e.str_val)

        elif cmd == 'MSET':
            for i in range(1, len(args) - 1, 2):
                e = s.reset(args[i]); e.type = 'string'; e.str_val = args[i+1]
            self._simple(conn, 'OK')

        elif cmd == 'MGET':
            self._w(conn, f'*{len(args)-1}\r\n')
            for k in args[1:]:
                e = s.get(k)
                if e and e.type == 'string': self._bulk(conn, e.str_val)
                else: self._nil(conn)

        elif cmd in ('INCR', 'INCRBY'):
            by = int(args[2]) if cmd == 'INCRBY' else 1
            e = s.make(args[1], 'string')
            if e.type != 'string': self._err(conn, WRONGTYPE); return
            if not e.str_val: e.str_val = '0'
            try: v = int(e.str_val) + by
            except ValueError:
                self._err(conn, 'ERR value is not an integer or out of range'); return
            e.str_val = str(v); self._int(conn, v)

        elif cmd == 'APPEND':
            e = s.make(args[1], 'string')
            if e.type != 'string': self._err(conn, WRONGTYPE); return
            e.str_val += args[2]; self._int(conn, len(e.str_val))

        elif cmd in ('LPUSH', 'RPUSH'):
            e = s.make(args[1], 'list')
            if e.type != 'list': self._err(conn, WRONGTYPE); return
            for v in args[2:]:
                if cmd == 'LPUSH': e.list_val.appendleft(v)
                else: e.list_val.append(v)
            self._int(conn, len(e.list_val))

        elif cmd in ('LPOP', 'RPOP'):
            e = s.get(args[1])
            if not e or not e.list_val: self._nil(conn); return
            if e.type != 'list': self._err(conn, WRONGTYPE); return
            left = cmd == 'LPOP'
            if len(args) >= 3:
                cnt = int(args[2]); res = []
                while cnt > 0 and e.list_val:
                    res.append(e.list_val.popleft() if left else e.list_val.pop()); cnt -= 1
                if not e.list_val: s.erase(args[1])
                self._arr(conn, res)
            else:
                v = e.list_val.popleft() if left else e.list_val.pop()
                if not e.list_val: s.erase(args[1])
                self._bulk(conn, v)

        elif cmd == 'LRANGE':
            e = s.get(args[1])
            if not e: self._earr(conn); return
            if e.type != 'list': self._err(conn, WRONGTYPE); return
            lst = list(e.list_val); ln = len(lst)
            a, b = int(args[2]), int(args[3])
            if a < 0: a = max(0, ln + a)
            if b < 0: b = ln + b
            b = min(b, ln - 1)
            if a > b: self._earr(conn); return
            self._arr(conn, lst[a:b+1])

        elif cmd == 'LLEN':
            e = s.get(args[1])
            if not e: self._int(conn, 0); return
            if e.type != 'list': self._err(conn, WRONGTYPE); return
            self._int(conn, len(e.list_val))

        elif cmd == 'SADD':
            e = s.make(args[1], 'set')
            if e.type != 'set': self._err(conn, WRONGTYPE); return
            n = 0
            for v in args[2:]:
                if v not in e.set_val: e.set_val.add(v); n += 1
            self._int(conn, n)

        elif cmd == 'SREM':
            e = s.get(args[1])
            if not e: self._int(conn, 0); return
            if e.type != 'set': self._err(conn, WRONGTYPE); return
            n = 0
            for v in args[2:]:
                if v in e.set_val: e.set_val.discard(v); n += 1
            if not e.set_val: s.erase(args[1])
            self._int(conn, n)

        elif cmd == 'SMEMBERS':
            e = s.get(args[1])
            if not e: self._earr(conn); return
            if e.type != 'set': self._err(conn, WRONGTYPE); return
            self._arr(conn, sorted(e.set_val))

        elif cmd == 'SISMEMBER':
            e = s.get(args[1])
            if not e: self._int(conn, 0); return
            if e.type != 'set': self._err(conn, WRONGTYPE); return
            self._int(conn, 1 if args[2] in e.set_val else 0)

        elif cmd == 'SCARD':
            e = s.get(args[1])
            if not e: self._int(conn, 0); return
            if e.type != 'set': self._err(conn, WRONGTYPE); return
            self._int(conn, len(e.set_val))

        elif cmd == 'ZADD':
            e = s.make(args[1], 'zset')
            if e.type != 'zset': self._err(conn, WRONGTYPE); return
            n = 0
            for i in range(2, len(args) - 1, 2):
                m = args[i+1]
                if m not in e.zset_val: n += 1
                e.zset_val[m] = float(args[i])
            self._int(conn, n)

        elif cmd == 'ZREM':
            e = s.get(args[1])
            if not e: self._int(conn, 0); return
            if e.type != 'zset': self._err(conn, WRONGTYPE); return
            n = 0
            for v in args[2:]:
                if v in e.zset_val: del e.zset_val[v]; n += 1
            if not e.zset_val: s.erase(args[1])
            self._int(conn, n)

        elif cmd == 'ZSCORE':
            e = s.get(args[1])
            if not e or args[2] not in e.zset_val: self._nil(conn); return
            self._bulk(conn, '%g' % e.zset_val[args[2]])

        elif cmd == 'ZRANK':
            e = s.get(args[1])
            if not e: self._nil(conn); return
            if e.type != 'zset': self._err(conn, WRONGTYPE); return
            srt = sorted(e.zset_val.items(), key=lambda x: (x[1], x[0]))
            for i, (m, _) in enumerate(srt):
                if m == args[2]: self._int(conn, i); return
            self._nil(conn)

        elif cmd == 'ZRANGE':
            e = s.get(args[1])
            if not e: self._earr(conn); return
            if e.type != 'zset': self._err(conn, WRONGTYPE); return
            srt = sorted(e.zset_val.items(), key=lambda x: (x[1], x[0]))
            ln = len(srt); a, b = int(args[2]), int(args[3])
            if a < 0: a = max(0, ln + a)
            if b < 0: b = ln + b
            b = min(b, ln - 1)
            ws = len(args) >= 5 and args[4].upper() == 'WITHSCORES'
            res = []
            for i in range(a, b + 1):
                res.append(srt[i][0])
                if ws: res.append('%g' % srt[i][1])
            if not res: self._earr(conn)
            else: self._arr(conn, res)

        elif cmd == 'ZCARD':
            e = s.get(args[1])
            if not e: self._int(conn, 0); return
            if e.type != 'zset': self._err(conn, WRONGTYPE); return
            self._int(conn, len(e.zset_val))

        else:
            self._err(conn, f"ERR unknown command '{args[0]}'")

    # ── Server loop ───────────────────────────────────────────────────────────

    def start(self, port):
        srv = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
        srv.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
        srv.bind(('', port))
        srv.listen(5)
        while True:
            conn, _ = srv.accept()
            buf = bytearray()
            while True:
                args = self._read_cmd(buf, conn)
                if args is None: break
                self._dispatch(conn, args)
            conn.close()

File: problems/problemset/mini-redis/Rust/solution/mini_redis.rs
use std::collections::{HashMap, VecDeque, BTreeSet};
use std::io::{BufRead, BufReader, Write};
use std::net::{TcpListener, TcpStream};

const WRONGTYPE: &str = "WRONGTYPE Operation against a key holding the wrong kind of value";

#[derive(Clone, PartialEq)]
enum EType { Str, List, Set, ZSet }

struct Entry {
    etype:     EType,
    str_val:   String,
    list_val:  VecDeque<String>,
    set_val:   BTreeSet<String>,
    zset_val:  HashMap<String, f64>,
    expire_at: i64,
}

impl Entry {
    fn new(t: EType) -> Self {
        Entry { etype: t, str_val: String::new(), list_val: VecDeque::new(),
                set_val: BTreeSet::new(), zset_val: HashMap::new(), expire_at: -1 }
    }
}

struct Store {
    data: HashMap<String, Entry>,
    now:  i64,
}

impl Store {
    fn new() -> Self { Store { data: HashMap::new(), now: 0 } }

    fn dead(&self, k: &str) -> bool {
        match self.data.get(k) {
            None    => true,
            Some(e) => e.expire_at >= 0 && e.expire_at <= self.now,
        }
    }

    fn alive(&mut self, k: &str) -> bool {
        if self.dead(k) { self.data.remove(k); false } else { true }
    }

    fn get(&mut self, k: &str) -> Option<&Entry> {
        if self.dead(k) { self.data.remove(k); None } else { self.data.get(k) }
    }

    fn get_mut(&mut self, k: &str) -> Option<&mut Entry> {
        if self.dead(k) { self.data.remove(k); None } else { self.data.get_mut(k) }
    }

    fn make(&mut self, k: &str, t: EType) -> &mut Entry {
        if self.dead(k) { self.data.remove(k); self.data.insert(k.to_string(), Entry::new(t)); }
        else if !self.data.contains_key(k) { self.data.insert(k.to_string(), Entry::new(t)); }
        self.data.get_mut(k).unwrap()
    }

    fn reset(&mut self, k: &str) -> &mut Entry {
        self.data.insert(k.to_string(), Entry::new(EType::Str));
        self.data.get_mut(k).unwrap()
    }

    fn erase(&mut self, k: &str) { self.data.remove(k); }
    fn flush(&mut self) { self.data.clear(); }
    fn tick(&mut self, n: i64) { self.now += n; }
    fn time(&self) -> i64 { self.now }

    fn dbsize(&mut self) -> i64 {
        let keys: Vec<String> = self.data.keys().cloned().collect();
        keys.iter().filter(|k| self.alive(k)).count() as i64
    }
}

// ── RESP helpers ──────────────────────────────────────────────────────────────

fn read_cmd(r: &mut BufReader<TcpStream>) -> Option<Vec<String>> {
    let mut line = String::new();
    r.read_line(&mut line).ok()?;
    let line = line.trim_end_matches(|c| c == '\r' || c == '\n');
    if !line.starts_with('*') { return None; }
    let argc: usize = line[1..].parse().ok()?;
    let mut args = Vec::with_capacity(argc);
    for _ in 0..argc {
        let mut blen = String::new();
        r.read_line(&mut blen).ok()?;
        let blen = blen.trim_end_matches(|c| c == '\r' || c == '\n');
        if !blen.starts_with('$') { return None; }
        let n: usize = blen[1..].parse().ok()?;
        let mut buf = vec![0u8; n + 2];
        use std::io::Read;
        r.read_exact(&mut buf).ok()?;
        args.push(String::from_utf8_lossy(&buf[..n]).to_string());
    }
    Some(args)
}

fn w(c: &mut TcpStream, s: &str)          { c.write_all(s.as_bytes()).ok(); }
fn simple(c: &mut TcpStream, s: &str)     { w(c, &format!("+{}\r\n", s)); }
fn rerr(c: &mut TcpStream, s: &str)       { w(c, &format!("-{}\r\n", s)); }
fn rint(c: &mut TcpStream, n: i64)        { w(c, &format!(":{}\r\n", n)); }
fn rnil(c: &mut TcpStream)                { w(c, "$-1\r\n"); }
fn earr(c: &mut TcpStream)                { w(c, "*0\r\n"); }
fn bulk(c: &mut TcpStream, s: &str)       { w(c, &format!("${}\r\n{}\r\n", s.len(), s)); }
fn arr(c: &mut TcpStream, items: &[String]) {
    w(c, &format!("*{}\r\n", items.len()));
    for item in items { bulk(c, item); }
}

fn fmt_g(f: f64) -> String {
    if f.fract() == 0.0 && f.abs() < 1e15 { format!("{}", f as i64) }
    else { format!("{}", f) }
}

// ── Dispatch ─────────────────────────────────────────────────────────────────

fn dispatch(s: &mut Store, c: &mut TcpStream, args: &[String]) {
    if args.is_empty() { return; }
    let cmd = args[0].to_uppercase();
    let cmd = cmd.as_str();

    match cmd {
        "PING" => {
            if args.len() > 1 { bulk(c, &args[1]); } else { simple(c, "PONG"); }
        }
        "FLUSHALL" => { s.flush(); simple(c, "OK"); }
        "DBSIZE"   => { rint(c, s.dbsize()); }
        "DEL" => {
            let n: i64 = args[1..].iter().map(|k| { if s.alive(k) { s.erase(k); 1 } else { 0 } }).sum();
            rint(c, n);
        }
        "EXISTS" => { rint(c, args[1..].iter().filter(|k| s.alive(k)).count() as i64); }
        "TYPE" => {
            let t = s.get(&args[1]).map(|e| match e.etype {
                EType::Str => "string", EType::List => "list",
                EType::Set => "set",   EType::ZSet => "zset",
            }).unwrap_or("none");
            simple(c, t);
        }
        "EXPIRE" => {
            let now = s.now;
            match s.get_mut(&args[1]) {
                None => rint(c, 0),
                Some(e) => { e.expire_at = now + args[2].parse::<i64>().unwrap_or(0); rint(c, 1); }
            }
        }
        "TTL" => {
            match s.get(&args[1]) {
                None    => rint(c, -2),
                Some(e) => rint(c, if e.expire_at < 0 { -1 } else { e.expire_at - s.now }),
            }
        }
        "DEBUG" => {
            if args.get(1).map(|s| s.to_uppercase()) == Some("TICK".into()) {
                if let Some(n) = args.get(2).and_then(|s| s.parse::<i64>().ok()) {
                    s.tick(n); simple(c, "OK");
                } else { rerr(c, "ERR"); }
            } else { rerr(c, "ERR unknown DEBUG subcommand"); }
        }
        "SET" => {
            let now = s.now;
            let e = s.reset(&args[1]); e.etype = EType::Str; e.str_val = args[2].clone();
            let mut i = 3;
            while i + 1 < args.len() {
                if args[i].to_uppercase() == "EX" {
                    if let Ok(n) = args[i+1].parse::<i64>() { e.expire_at = now + n; }
                    i += 2;
                } else { i += 1; }
            }
            simple(c, "OK");
        }
        "GET" => {
            match s.get(&args[1]) {
                None => rnil(c),
                Some(e) if e.etype != EType::Str => rerr(c, WRONGTYPE),
                Some(e) => { let v = e.str_val.clone(); bulk(c, &v); }
            }
        }
        "MSET" => {
            let mut i = 1;
            while i + 1 < args.len() {
                let e = s.reset(&args[i]); e.etype = EType::Str; e.str_val = args[i+1].clone();
                i += 2;
            }
            simple(c, "OK");
        }
        "MGET" => {
            w(c, &format!("*{}\r\n", args.len() - 1));
            for k in &args[1..] {
                match s.get(k) {
                    Some(e) if e.etype == EType::Str => { let v = e.str_val.clone(); bulk(c, &v); }
                    _ => rnil(c),
                }
            }
        }
        "INCR" | "INCRBY" => {
            let by: i64 = if cmd == "INCRBY" { args[2].parse().unwrap_or(0) } else { 1 };
            let e = s.make(&args[1], EType::Str);
            if e.etype != EType::Str { rerr(c, WRONGTYPE); return; }
            if e.str_val.is_empty() { e.str_val = "0".to_string(); }
            match e.str_val.parse::<i64>() {
                Err(_) => { rerr(c, "ERR value is not an integer or out of range"); }
                Ok(v)  => { let nv = v + by; e.str_val = nv.to_string(); rint(c, nv); }
            }
        }
        "APPEND" => {
            let e = s.make(&args[1], EType::Str);
            if e.etype != EType::Str { rerr(c, WRONGTYPE); return; }
            e.str_val.push_str(&args[2]);
            rint(c, e.str_val.len() as i64);
        }
        "LPUSH" | "RPUSH" => {
            let e = s.make(&args[1], EType::List);
            if e.etype != EType::List { rerr(c, WRONGTYPE); return; }
            for v in &args[2..] {
                if cmd == "LPUSH" { e.list_val.push_front(v.clone()); }
                else              { e.list_val.push_back(v.clone()); }
            }
            rint(c, e.list_val.len() as i64);
        }
        "LPOP" | "RPOP" => {
            match s.get_mut(&args[1]) {
                None => rnil(c),
                Some(e) if e.list_val.is_empty() => rnil(c),
                Some(e) if e.etype != EType::List => rerr(c, WRONGTYPE),
                Some(e) => {
                    let left = cmd == "LPOP";
                    if args.len() >= 3 {
                        let mut cnt: usize = args[2].parse().unwrap_or(0);
                        let mut res = Vec::new();
                        while cnt > 0 && !e.list_val.is_empty() {
                            let v = if left { e.list_val.pop_front() } else { e.list_val.pop_back() };
                            res.push(v.unwrap()); cnt -= 1;
                        }
                        let empty = e.list_val.is_empty();
                        arr(c, &res);
                        if empty { s.erase(&args[1]); }
                    } else {
                        let v = if left { e.list_val.pop_front() } else { e.list_val.pop_back() };
                        let v = v.unwrap();
                        let empty = e.list_val.is_empty();
                        bulk(c, &v);
                        if empty { s.erase(&args[1]); }
                    }
                }
            }
        }
        "LRANGE" => {
            match s.get(&args[1]) {
                None => earr(c),
                Some(e) if e.etype != EType::List => rerr(c, WRONGTYPE),
                Some(e) => {
                    let lst: Vec<String> = e.list_val.iter().cloned().collect();
                    let ln = lst.len() as i64;
                    let mut a = args[2].parse::<i64>().unwrap_or(0);
                    let mut b = args[3].parse::<i64>().unwrap_or(-1);
                    if a < 0 { a = (ln + a).max(0); }
                    if b < 0 { b = ln + b; }
                    b = b.min(ln - 1);
                    if a > b { earr(c); return; }
                    let res: Vec<String> = lst[a as usize ..= b as usize].to_vec();
                    arr(c, &res);
                }
            }
        }
        "LLEN" => {
            match s.get(&args[1]) {
                None    => rint(c, 0),
                Some(e) if e.etype != EType::List => rerr(c, WRONGTYPE),
                Some(e) => rint(c, e.list_val.len() as i64),
            }
        }
        "SADD" => {
            let e = s.make(&args[1], EType::Set);
            if e.etype != EType::Set { rerr(c, WRONGTYPE); return; }
            let n: i64 = args[2..].iter().map(|v| if e.set_val.insert(v.clone()) { 1 } else { 0 }).sum();
            rint(c, n);
        }
        "SREM" => {
            match s.get_mut(&args[1]) {
                None    => rint(c, 0),
                Some(e) if e.etype != EType::Set => rerr(c, WRONGTYPE),
                Some(e) => {
                    let n: i64 = args[2..].iter().map(|v| if e.set_val.remove(v.as_str()) { 1 } else { 0 }).sum();
                    let empty = e.set_val.is_empty();
                    rint(c, n);
                    if empty { s.erase(&args[1]); }
                }
            }
        }
        "SMEMBERS" => {
            match s.get(&args[1]) {
                None    => earr(c),
                Some(e) if e.etype != EType::Set => rerr(c, WRONGTYPE),
                Some(e) => { let v: Vec<String> = e.set_val.iter().cloned().collect(); arr(c, &v); }
            }
        }
        "SISMEMBER" => {
            match s.get(&args[1]) {
                None    => rint(c, 0),
                Some(e) if e.etype != EType::Set => rerr(c, WRONGTYPE),
                Some(e) => rint(c, if e.set_val.contains(&args[2]) { 1 } else { 0 }),
            }
        }
        "SCARD" => {
            match s.get(&args[1]) {
                None    => rint(c, 0),
                Some(e) if e.etype != EType::Set => rerr(c, WRONGTYPE),
                Some(e) => rint(c, e.set_val.len() as i64),
            }
        }
        "ZADD" => {
            let e = s.make(&args[1], EType::ZSet);
            if e.etype != EType::ZSet { rerr(c, WRONGTYPE); return; }
            let mut n = 0i64;
            let mut i = 2;
            while i + 1 < args.len() {
                let score: f64 = args[i].parse().unwrap_or(0.0);
                let member = args[i+1].clone();
                if !e.zset_val.contains_key(&member) { n += 1; }
                e.zset_val.insert(member, score);
                i += 2;
            }
            rint(c, n);
        }
        "ZREM" => {
            match s.get_mut(&args[1]) {
                None    => rint(c, 0),
                Some(e) if e.etype != EType::ZSet => rerr(c, WRONGTYPE),
                Some(e) => {
                    let n: i64 = args[2..].iter().map(|v| if e.zset_val.remove(v.as_str()).is_some() { 1 } else { 0 }).sum();
                    let empty = e.zset_val.is_empty();
                    rint(c, n);
                    if empty { s.erase(&args[1]); }
                }
            }
        }
        "ZSCORE" => {
            match s.get(&args[1]) {
                None    => rnil(c),
                Some(e) => match e.zset_val.get(&args[2]) {
                    None    => rnil(c),
                    Some(&sc) => { let fs = fmt_g(sc); bulk(c, &fs); }
                }
            }
        }
        "ZRANK" => {
            match s.get(&args[1]) {
                None    => rnil(c),
                Some(e) if e.etype != EType::ZSet => rerr(c, WRONGTYPE),
                Some(e) => {
                    let mut srt: Vec<(f64, &String)> = e.zset_val.iter().map(|(m, &sc)| (sc, m)).collect();
                    srt.sort_by(|a, b| a.0.partial_cmp(&b.0).unwrap().then(a.1.cmp(b.1)));
                    match srt.iter().position(|(_, m)| *m == &args[2]) {
                        None    => rnil(c),
                        Some(i) => rint(c, i as i64),
                    }
                }
            }
        }
        "ZRANGE" => {
            match s.get(&args[1]) {
                None    => earr(c),
                Some(e) if e.etype != EType::ZSet => rerr(c, WRONGTYPE),
                Some(e) => {
                    let mut srt: Vec<(f64, &String)> = e.zset_val.iter().map(|(m, &sc)| (sc, m)).collect();
                    srt.sort_by(|a, b| a.0.partial_cmp(&b.0).unwrap().then(a.1.cmp(b.1)));
                    let ln = srt.len() as i64;
                    let mut a = args[2].parse::<i64>().unwrap_or(0);
                    let mut b = args[3].parse::<i64>().unwrap_or(-1);
                    if a < 0 { a = (ln + a).max(0); }
                    if b < 0 { b = ln + b; }
                    b = b.min(ln - 1);
                    let ws = args.get(4).map(|s| s.to_uppercase() == "WITHSCORES").unwrap_or(false);
                    let mut res = Vec::new();
                    for i in a..=b {
                        res.push(srt[i as usize].1.clone());
                        if ws { res.push(fmt_g(srt[i as usize].0)); }
                    }
                    if res.is_empty() { earr(c); } else { arr(c, &res); }
                }
            }
        }
        "ZCARD" => {
            match s.get(&args[1]) {
                None    => rint(c, 0),
                Some(e) if e.etype != EType::ZSet => rerr(c, WRONGTYPE),
                Some(e) => rint(c, e.zset_val.len() as i64),
            }
        }
        _ => { rerr(c, &format!("ERR unknown command '{}'", args[0])); }
    }
}

// ── Server ────────────────────────────────────────────────────────────────────

pub struct MiniRedis {
    store: Store,
}

impl MiniRedis {
    pub fn new() -> Self { MiniRedis { store: Store::new() } }

    pub fn start(&mut self, port: u16) {
        let listener = TcpListener::bind(("0.0.0.0", port)).unwrap();
        for stream in listener.incoming() {
            let Ok(stream) = stream else { continue };
            let reader_stream = stream.try_clone().unwrap();
            let mut writer = stream;
            let mut reader = BufReader::new(reader_stream);
            loop {
                match read_cmd(&mut reader) {
                    None    => break,
                    Some(args) => dispatch(&mut self.store, &mut writer, &args),
                }
            }
        }
    }
}
